### ISSUE TITLE
Race condition in the `Timed` trait implementation

### ISSUE THREAD

**Issue by axel22** [scalatest/scalatest#1147]
This has been biting me for a while now, so I decided to dig in:

```
[info] - reactor should ensure the ReactorStarted event *** FAILED ***
[info]   java.lang.IllegalStateException: Timer already cancelled.
[info]   at java.util.Timer.sched(Timer.java:397)
[info]   at java.util.Timer.schedule(Timer.java:193)
[info]   at org.scalatest.Timer.schedule(JavaClassesWrappers.scala:93)
[info]   at org.scalatest.enablers.Timed$$anon$3.timeoutAfter(Timed.scala:335)
[info]   at org.scalatest.enablers.Timed$$anon$3.timeoutAfter(Timed.scala:286)
[info]   at org.scalatest.concurrent.TimeLimits.failAfterImpl(TimeLimits.scala:239)
[info]   at org.scalatest.concurrent.TimeLimits.failAfterImpl$(TimeLimits.scala:233)
[info]   at io.reactors.concurrent.ReactorSystemTest.failAfterImpl(reactor-system-tests.scala:206)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:230)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:229)
```

I believe that this is a race between the future created in line 306 here, which calls cancel in line 310:

https://github.com/scalatest/scalatest/blob/3.1.x/scalatest/src/main/scala/org/scalatest/enablers/Timed.scala#L310

and the `schedule` call on the same timer in line 335:

https://github.com/scalatest/scalatest/blob/3.1.x/scalatest/src/main/scala/org/scalatest/enablers/Timed.scala#L335

A similar race condition tends to happen with the line 309, where `task.cancel` is called.
Usually, this happens when the asynchronous test finishes very quickly, enabling it to *race* with the ScalaTest code.

The solution could be to add another promise in line 305, and use it to atomically decide on whether the task is canceled or scheduled:

```
val timerOwner = Promise[Unit]()
```

and then in line 309:

```
if (timerOwner.trySuccess(())) {
  task.cancel()
  timer.cancel()
}
```

and in line 335:

```
if (timerOwner.trySuccess(())) {
  timer.schedule(task, delay)
}
```

**Comment by axel22** [comment]
@bvenners Do you agree with this analysis? Please let me know, and I will send a PR with a fix.

**Comment by axel22** [comment]
@cheeseng Thoughts?

**Comment by cheeseng** [comment]
@axel22 let me take a look now.

**Comment by cheeseng** [comment]
@axel22 I wonder if it would work if we move the schedule call before the future creation, what do you think?

**Comment by axel22** [comment]
As far as I can see, that would also solve the problem. A comment saying that the order of calling `schedule` should not be changed would make it future-proof.

**Comment by cheeseng** [comment]
@axel22 I submitted a PR for this: 

https://github.com/scalatest/scalatest/pull/1155

Hopefully @bvenners will be able to review soon.

Thanks for your great analysis!

**Comment by axel22** [comment]
Awesome!

**Comment by bvenners** [comment]
#1155 is merged into 3.0.x and cherry-picked into 3.1.x and feature-algebra-and-laws-2

**Comment by axel22** [comment]
Thanks a lot! Looking forward to the release!

**Comment by cheeseng** [comment]
@axel22 the just released 3.0.4 includes this fix.

**Comment by axel22** [comment]
That is really great - thank you for letting me know!
